File: archive/cdk-go/Pr2307/lib/constructs/storage_construct.go
package constructs

import (
	"github.com/aws/aws-cdk-go/awscdk/v2"
	"github.com/aws/aws-cdk-go/awscdk/v2/awss3"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type StorageConstructProps struct {
	Environment string
}

type StorageConstruct struct {
	constructs.Construct
	Bucket        awss3.IBucket
	LoggingBucket awss3.IBucket
}

func NewStorageConstruct(scope constructs.Construct, id string, props *StorageConstructProps) *StorageConstruct {
	construct := constructs.NewConstruct(scope, &id)

	// Enhanced logging bucket
	loggingBucket := awss3.NewBucket(construct, jsii.String("AccessLogsBucket"), &awss3.BucketProps{
		BucketName:        jsii.String("proj-s3-logs-" + props.Environment),
		PublicReadAccess:  jsii.Bool(false),
		BlockPublicAccess: awss3.BlockPublicAccess_BLOCK_ALL(),
		Encryption:        awss3.BucketEncryption_S3_MANAGED,
		EnforceSSL:        jsii.Bool(true),
		LifecycleRules: &[]*awss3.LifecycleRule{
			{
				Id:         jsii.String("DeleteOldAccessLogs"),
				Enabled:    jsii.Bool(true),
				Expiration: awscdk.Duration_Days(jsii.Number(90)),
				Transitions: &[]*awss3.Transition{
					{
						StorageClass:    awss3.StorageClass_INFREQUENT_ACCESS(),
						TransitionAfter: awscdk.Duration_Days(jsii.Number(30)),
					},
				},
			},
		},
	})

	// Enhanced main bucket with SSL-only policy and Transfer Acceleration
	bucket := awss3.NewBucket(construct, jsii.String("MainBucket"), &awss3.BucketProps{
		BucketName:             jsii.String("proj-s3-" + props.Environment),
		Versioned:              jsii.Bool(true),
		PublicReadAccess:       jsii.Bool(false),
		BlockPublicAccess:      awss3.BlockPublicAccess_BLOCK_ALL(),
		Encryption:             awss3.BucketEncryption_S3_MANAGED,
		EnforceSSL:             jsii.Bool(true),
		TransferAcceleration:   jsii.Bool(true),
		ServerAccessLogsBucket: loggingBucket,
		ServerAccessLogsPrefix: jsii.String("access-logs/"),
		EventBridgeEnabled:     jsii.Bool(true),
		// Enhanced lifecycle rules for cost optimization
		LifecycleRules: &[]*awss3.LifecycleRule{
			{
				Id:      jsii.String("CostOptimizationRule"),
				Enabled: jsii.Bool(true),
				Transitions: &[]*awss3.Transition{
					{
						StorageClass:    awss3.StorageClass_INFREQUENT_ACCESS(),
						TransitionAfter: awscdk.Duration_Days(jsii.Number(30)),
					},
					{
						StorageClass:    awss3.StorageClass_GLACIER(),
						TransitionAfter: awscdk.Duration_Days(jsii.Number(90)),
					},
					{
						StorageClass:    awss3.StorageClass_DEEP_ARCHIVE(),
						TransitionAfter: awscdk.Duration_Days(jsii.Number(365)),
					},
				},
				// Clean up old versions and incomplete multipart uploads
				NoncurrentVersionTransitions: &[]*awss3.NoncurrentVersionTransition{
					{
						StorageClass:               awss3.StorageClass_INFREQUENT_ACCESS(),
						TransitionAfter:            awscdk.Duration_Days(jsii.Number(30)),
						NoncurrentVersionsToRetain: jsii.Number(3),
					},
				},
				NoncurrentVersionExpiration:         awscdk.Duration_Days(jsii.Number(100)),
				AbortIncompleteMultipartUploadAfter: awscdk.Duration_Days(jsii.Number(7)),
			},
		},
	})

	// Note: SSL-only policy configuration simplified for CDK Go compatibility
	// The EnforceSSL: true property above provides the same security benefit

	// Note: IntelligentTiering configuration is not directly supported in CDK Go
	// Consider using lifecycle rules above for cost optimization or configure
	// IntelligentTiering manually in the AWS Console after deployment

	// Add resource tags for better management
	awscdk.Tags_Of(bucket).Add(jsii.String("Environment"), jsii.String(props.Environment), nil)
	awscdk.Tags_Of(bucket).Add(jsii.String("Project"), jsii.String("tap-infrastructure"), nil)
	awscdk.Tags_Of(bucket).Add(jsii.String("BackupEnabled"), jsii.String("true"), nil)
	awscdk.Tags_Of(bucket).Add(jsii.String("CostOptimized"), jsii.String("true"), nil)

	awscdk.Tags_Of(loggingBucket).Add(jsii.String("Environment"), jsii.String(props.Environment), nil)
	awscdk.Tags_Of(loggingBucket).Add(jsii.String("Project"), jsii.String("tap-infrastructure"), nil)
	awscdk.Tags_Of(loggingBucket).Add(jsii.String("Purpose"), jsii.String("access-logs"), nil)

	return &StorageConstruct{
		Construct:     construct,
		Bucket:        bucket,
		LoggingBucket: loggingBucket,
	}
}

File: archive/cdk-ts/Pr5998/lib/constructs/storage-construct.ts
import * as cdk from 'aws-cdk-lib';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export interface StorageConstructProps {
  environmentSuffix: string;
  retentionDays: number;
}

export class StorageConstruct extends Construct {
  public readonly transactionBucket: s3.Bucket;

  constructor(scope: Construct, id: string, props: StorageConstructProps) {
    super(scope, id);

    const { environmentSuffix, retentionDays } = props;

    // Create S3 bucket for transaction logs with Intelligent-Tiering
    // This works with ALL retention periods including 7 days for dev
    this.transactionBucket = new s3.Bucket(this, 'TransactionBucket', {
      bucketName: `payment-transactions-${environmentSuffix}`,
      versioned: true,
      encryption: s3.BucketEncryption.S3_MANAGED,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      enforceSSL: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
      lifecycleRules: [
        {
          id: `intelligent-tiering-${environmentSuffix}`,
          enabled: true,
          // Use Intelligent-Tiering storage class - works with any retention period
          transitions: [
            {
              storageClass: s3.StorageClass.INTELLIGENT_TIERING,
              transitionAfter: cdk.Duration.days(0),
            },
          ],
        },
        {
          id: `expiration-${environmentSuffix}`,
          enabled: true,
          // Expire objects after retention period
          expiration: cdk.Duration.days(retentionDays),
        },
        {
          id: `abort-incomplete-multipart-upload-${environmentSuffix}`,
          enabled: true,
          // Clean up incomplete multipart uploads after 7 days
          abortIncompleteMultipartUploadAfter: cdk.Duration.days(7),
        },
      ],
    });

    // Tags
    cdk.Tags.of(this.transactionBucket).add(
      'Name',
      `payment-transactions-${environmentSuffix}`
    );
    cdk.Tags.of(this.transactionBucket).add('Environment', environmentSuffix);
  }
}
